### PR TITLE
feat: search with algolia

### LINKED ISSUE #69: Improve module search
Successor to #56

## Problem

1. The search results are not biased by star count. Some examples:
   - term: _discord_: the highest star count module is only on place 6.
2. There are no filtering or alternative sorting capabilities (last uploaded at, supports typescript, no external imports)
3. Search is not always very relevant to what was searched.
4. No filtering by module tags

## Solution

1. We need to include more data in the documents being searched (github tags, last upload date, and other stats/metrics).
2. Use a better search engine. I would like to use MeiliSearch because of the search customizability and API (cc @erlend-sh). We need to find a way to host it / get access to a hosted instance. This would also involve pushing updates to the documents in the database to the search engine. This can be accomplished relatively easily.

### PR DESCRIPTION
**Pull request by lucacasonato**
Closes #69

This adds search via Algolia. We attempted to use Meilisearch previously, but automatically setting up hosting was too﻿ complex.

This is untestable locally, because there is no local Algolia emulator.

### CODE REVIEW

File: api/async/stargazers.ts [hunk 1]
```diff
@@ -44,6 +50,7 @@ export async function handler(
       if (repo.stargazers_count !== module.star_count) {
         module.star_count = repo.stargazers_count;
         await database.saveModule(module as unknown as Module);
+        await algolia.saveDatabaseModule(module);
```
Comment (wperron) [inline api/async/stargazers.ts:53]: probably better to `Promise.all` both operations

File: api/webhook/github.ts [hunk 2]
```diff
@@ -152,20 +157,24 @@ async function pingEvent(
...
     repo_id: repoId,
     owner,
     repo,
     description,
+    // deno-lint-ignore camelcase
     star_count: starCount,
-  });
+  };
+  await database.saveModule(module);
+  await algolia.saveDatabaseModule(module);
```
Comment (wperron) [inline api/webhook/github.ts:177]: same here

File: api/webhook/github.ts [hunk 3]
```diff
@@ -347,20 +356,24 @@ async function initiateBuild(
...
```
Comment (wperron) [inline api/webhook/github.ts:376]: and here

**Comment by wperron** [comment]
About the deprecation header; what's the strategy to migrate the search over to algolia? is the `/modules?query` simply going to not work after May 1st?

**Comment by lucacasonato** [comment]
> About the deprecation header; what's the strategy to migrate the search over to algolia? is the `/modules?query` simply going to not work after May 1st?

Yes. The website will move over to query Algolia directly.

**Comment by wperron** [comment]
> > About the deprecation header; what's the strategy to migrate the search over to algolia? is the `/modules?query` simply going to not work after May 1st?
> 
> Yes. The website will move over to query Algolia directly. 

There's also the option to redirect to the new search endpoint during the transition instead of just returning an error

**Comment by lucacasonato** [comment]
The new search endpoint won't be for public use though.

**Review by wperron — APPROVED** [review]

**Comment by ry** [comment]
@lucacasonato Please rebase or close

**Comment by ry** [comment]
stale

**Pull request closed without merging**